### PR TITLE
build: add a CMake based build system

### PR DESCRIPTION
**Pull request by compnerd**
This will allow using Swift System in the toolchain stack, allowing us
to bootstrap a toolchain.

### CODE REVIEW

File: Sources/System/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  Errno.swift
+  FileDescriptor.swift
+  FileHelpers.swift
+  FileOperations.swift
+  FilePermissions.swift
+  PlatformString.swift
+  SystemString.swift
+  Util.swift
+  UtilConsumers.swift)
+target_sources(System PRIVATE
```
Comment (milseman) [inline Sources/System/CMakeLists.txt:20]: What does `PRIVATE` mean?

File: Sources/System/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+This source file is part of the Swift System open source project
+
+Copyright (c) 2019 Apple Inc. and the Swift Numerics project authors
+Licensed under Apache License v2.0 with Runtime Library Exception
+
+See https://swift.org/LICENSE.txt for license information
+#]]
+
+add_library(System
+  Errno.swift
```
Comment (milseman) [inline Sources/System/CMakeLists.txt:11]: I don't suppose there's any reasonable way to say `*.swift` within a  directory? SPM already dictates that our source code matches our targets

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#[[
+This source file is part of the Swift System open source project
+
+Copyright (c) 2019 Apple Inc. and the Swift Numerics project authors
+Licensed under Apache License v2.0 with Runtime Library Exception
+
+See https://swift.org/LICENSE.txt for license information
+#]]
```
Comment (milseman) [inline CMakeLists.txt:8]: Change comment to System's header comment.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+include(SwiftSupport)
+
+add_subdirectory(Sources)
+
+get_property(SWIFT_SYSTEM_EXPORTS GLOBAL PROPERTY SWIFT_SYSTEM_EXPORTS)
+export(TARGETS ${SWIFT_SYSTEM_EXPORTS}
+  NAMESPACE SwiftSystem::
+  FILE swift-system-config.cmake
+  EXPORT_LINK_INTERFACE_LIBRARIES)
```
Comment (milseman) [inline CMakeLists.txt:29]: What does this `export` do?

File: Sources/System/CMakeLists.txt (hunk 1 above)
Comment (compnerd) [reply Sources/System/CMakeLists.txt:20]: In the case of `target_sources` the `PRIVATE` indicates that the listed sources are not meant to be used to populate `INTERFACE_SOURCES` for the target.  This means that the sources listed here do not get propagated to users of this library.

File: Sources/System/CMakeLists.txt (hunk 2 above)
Comment (compnerd) [reply Sources/System/CMakeLists.txt:11]: No, that is the way that CMake is designed - to be explicit.  If you try to use a glob, any changes to the source tree structure will not be reflected in the build until a clean build is done.

File: CMakeLists.txt (hunk 3 above)
Comment (compnerd) [reply CMakeLists.txt:8]: Sure, that should be easy to do.

File: CMakeLists.txt (hunk 4 above)
Comment (compnerd) [reply CMakeLists.txt:29]: This export is the trick that enables wiring up the entire toolchain build together.  It creates a `swift-system-config.cmake` which can then be used for any subsequent build to just grab the build artifacts from the build tree without having to install and create a single distribution image and incrementally add content into it.  Once everything is built and verified to work together, you just create a single image that you intend to distribute.

File: Sources/System/CMakeLists.txt (hunk 1 above)
Comment (milseman) [reply Sources/System/CMakeLists.txt:20]: But do top-level sources? Unlike the `Internals` folder, this is just a way of organizing `FilePath` code together for navigation purposes. There's still API inside.

**Comment by milseman** [comment]
@swift-ci please test

File: Sources/System/CMakeLists.txt (hunk 1 above)
Comment (compnerd) [reply Sources/System/CMakeLists.txt:20]: ah, I see your point.  No, the top-level sources are treated as private by default.  I only did the split `target_sources` for reading.  We can even merge them into a single call if you prefer.  That is a stylistic thing.

Comment (milseman) [reply Sources/System/CMakeLists.txt:20]: I have no opinions here, just hoping to understand CMake a little bit better :-)

**Comment by compnerd** [comment]
@swift-ci please test

**Pull request merged**